File: 소스코드/xgboost.py
import pandas as pd
import numpy as np
from SyncRNG import SyncRNG
from xgboost import XGBClassifier
from sklearn.metrics import confusion_matrix
import os
os.getcwd()
os.chdir('C:/Google driver/pythonProject_study')

#데이터로드
raw_data = pd.read_csv('diabetes.csv')



# 데이터 셋  7:3 으로 분할
v=list(range(1,len(raw_data)+1))
s=SyncRNG(seed=42)
ord=s.shuffle(v)
idx=ord[:round(len(raw_data)*0.7)]

# R에서는 데이터프레임이 1부터 시작하기 때문에
# python에서 0행과 R에서 1행이 같은 원리로
# 같은 인덱스 번호를 가진다면 -1을 해주어 같은 데이터를 가지고 오게 한다.
# 인덱스 수정-R이랑 같은 데이터 가져오려고
for i in range(0,len(idx)):
    idx[i]=idx[i]-1

# 학습데이터, 테스트데이터 생성
train=raw_data.loc[idx] # 70%
test=raw_data.drop(idx) # 30%

#전처리

y_train=pd.DataFrame(train.Outcome)
x_train=pd.DataFrame(train.drop('Outcome',axis=1))
y_test=pd.DataFrame(test.Outcome)
x_test=pd.DataFrame(test.drop('Outcome',axis=1))

# xgboost
xgb_model  = XGBClassifier()
xgb_model .fit(x_train, y_train)

confusion_matrix(y_test, xgb_model.predict(x_test))
xgb_pred  = xgb_model .predict(x_test)

xgb_model .score(x_test, y_test)

# xgboost_tune 모델
xgb_tune_model = xgb.XGBClassifier(booster = 'gbtree',
                                   learning_rate = 0.02,
                                   n_estimators=1000,
                                   min_child_weight=3 ,
                                   max_depth=8)
xgb_tune_model.fit(x_train, y_train,
                   early_stopping_rounds=30,
                   eval_set=[(x_test, y_test)],
                   verbose=False )
confusion_matrix(y_test, xgb_tune_model.predict(x_test))
xgb_tune_pred = xgb_tune_model.predict(x_test)

from sklearn.metrics import classification_report
from sklearn import metrics
print(classification_report(y_test, xgb_pred ))
metrics.f1_score(y_test, xgb_pred )
print(classification_report(y_test, xgb_tune_pred))
metrics.f1_score(y_test, xgb_tune_pred)
#####################
# 0.6754966887417219

"""
R
# f1 score
caret::confusionMatrix(as.factor(test_y), as.factor(XGB_pred))$byClass[7]
#########
# 0.81
"""

File: 소스코드/xgboost.R
# install.packages("xgboost")
# install.packages("Matrix")
library(xgboost)
library(Matrix)

df<-read.csv("C:/Google driver/pythonProject_study/diabetes.csv")
df_xgboost <- df #원본데이터를 변경하지 않기 위해 복사
df

df_xgb_sparse_matrix <- sparse.model.matrix(Outcome~.-1,data=df_xgboost)

# train, test data set, label data 생성
library(SyncRNG) 
v <- 1:nrow(df)
s <- SyncRNG(seed=42)
idx <- s$shuffle(v)[1:round(nrow(df_xgb_sparse_matrix)*0.7)] 
idx[1:length(idx)]
# train <- df[idx,] 
# test <- df[-idx,]
train_x <- df_xgb_sparse_matrix[idx,]
test_x <- df_xgb_sparse_matrix[-idx,]
train_y <- df_xgboost[idx,'Outcome']
test_y <- df_xgboost[-idx,'Outcome']

# xgboost알고리즘 사용을 위한 데이터 형태 변환
dtrain <- xgb.DMatrix(data = train_x, label = as.matrix(train_y))
dtest <- xgb.DMatrix(data = test_x, label = as.matrix(test_y))



# xgb모델 생성
xgbmodel <- xgboost(data = dtrain,                     
                    max.depth=5,                            
                    nrounds=20)                             

# xgb모델 예측
XGB_pred <- predict(xgbmodel, dtest)
XGB_pred <- ifelse(XGB_pred >= 0.5, 1, 0)

# 해당 모델의 중요 변수
library(dplyr)
xgb.importance(colnames(dtrain), model = xgb) %>% 
  xgb.plot.importance(top_n = 30)


# 해당 모델 저장
xgb.save(xgb,"xgboost.model")



# caret 기본모델
caret_xgb_model<- train(Outcome ~ ., data = train,
                        method = "xgbTree", tuneGrid= caretxgbGrid)
caret_xgb_pred<-predict(caret_xgb_model, newdata = test)

# caret 튜닝모델
caretxgbGrid <-  expand.grid(nrounds = 200,
                             max_depth = 5,
                             eta = 0.05,
                             gamma = 0.01,
                             colsample_bytree = 0.75,
                             min_child_weight = 0,
                             subsample = 0.5)

caret_tune_xgb_model <- train(Outcome ~ ., data = train, 
                              method = "xgbTree", tuneGrid= caretxgbGrid)

caret_tune_xgb_pred<-predict(caret_tune_xgb_model, newdata = test)


# f1 score
caret::confusionMatrix(as.factor(test_y), as.factor(XGB_pred))$byClass #R 평가  #0.8407643
caret::confusionMatrix(caret_xgb_pred,test$Outcome)$byClass #caret평가 # 0.8246753
caret::confusionMatrix(caret_tune_xgb_pred,test$Outcome)$byClass #caret tune평가 #0.8424437
